### PR TITLE
Refactoring skill book use to allow for adjacent reading.

### PR DESCRIPTION
**Pull request by MistakeNot4892**
## Description of changes
You can now continue to reference a skill book so long as it is either in your direct inventory (hands, pocket, belt as applicable) or on a tile adjacent to you (put it on a table, etc).

## Why and what will this PR improve
This will make skillbooks more broadly useful for things like piloting or construction as they won't take up an entire hand slot permanently. It will also allow Baxxid to use skillbooks properly on downstream scav.

## Changelog
:cl:
tweak: Skill books can now be used from adjacent tiles.
/:cl:

### CODE REVIEW

File: code/game/objects/items/books/skill_book.dm [hunk 1]
```diff
@@ -82,64 +82,106 @@ Skill books that increase your skills while you activate and hold them
...
 	if(user.get_skill_value(skill) > skill_req)
 		to_chat(user, SPAN_WARNING("You already know everything [title] has to teach you!"))
-		return
+		return FALSE
+
+	return TRUE
+
+/obj/item/book/skill/attack_self(mob/user)
+
+	if(reading && reading == weakref(user))
```
Comment (afterthought2) [inline code/game/objects/items/books/skill_book.dm]: I believe the correct check is always `user && user == reading?.resolve()` or equivalent. Avoids creating unnecessary weakrefs and is independent of the weakref implementation.

File: code/game/objects/items/books/skill_book.dm [hunk 2]
```diff
@@ -82,64 +82,106 @@ Skill books that increase your skills while you activate and hold them
...
+		//Close book, get rid of buffs
+		src.unlearn(user)
```
Comment (afterthought2) [inline code/game/objects/items/books/skill_book.dm]: Extraneous `src`.

File: code/game/objects/items/books/skill_book.dm [hunk 3]
```diff
@@ -82,64 +82,106 @@ Skill books that increase your skills while you activate and hold them
...
+		to_chat(user, SPAN_NOTICE("You close the [name]. That's enough learning for now."))
```
Comment (afterthought2) [inline code/game/objects/items/books/skill_book.dm]: Probably `\the [name]`?

File: code/game/objects/items/books/skill_book.dm [hunk 4]
```diff
@@ -82,64 +82,106 @@ Skill books that increase your skills while you activate and hold them
...
+	if(!reading)
+		return PROCESS_KILL
+	check_buff()
+
+/obj/item/book/skill/proc/check_buff()
+	var/mob/R = reading?.resolve()
+	if(!istype(R) || R.incapacitated())
+		remove_buff()
+		return
+	if(loc != R && !R.Adjacent(src))
```
Comment (afterthought2) [inline code/game/objects/items/books/skill_book.dm]: I would at least consider `CanPhysicallyInteract` (possibly for both this and the incapacitated check above).

File: code/game/objects/items/books/skill_book.dm [hunk 5]
```diff
@@ -82,64 +82,106 @@ Skill books that increase your skills while you activate and hold them
...
-	if(reading)
-		to_chat(user, SPAN_DANGER("You lose the page you were on! You can't cross-reference using the [name] like this!"))
-		var/mob/M = user
-		if(istype(M) && M.fetch_buffs_of_type(/datum/skill_buff/skill_book, 0))
-			src.unlearn(user)
-		reading = FALSE
+/obj/item/book/skill/Move()
 	. = ..()
+	if(.)
+		check_buff()
```
Comment (afterthought2) [inline code/game/objects/items/books/skill_book.dm]: This override won't work with forceMove, if you e.g. store the book in a bag. Is this intentional? If not, consider moved events instead.

File: code/game/objects/items/books/skill_book.dm [hunk 6]
```diff
@@ -82,64 +82,106 @@ Skill books that increase your skills while you activate and hold them
...
+
+/obj/item/book/skill/dropped()
+	check_buff()
```
Comment (afterthought2) [inline code/game/objects/items/books/skill_book.dm]: That seems to answer above question with `forceMove`. I'd consider subscribing to your own moved event.

File: code/game/objects/items/books/skill_book.dm (hunk 3 above)
Comment (MistakeNot4892) [reply code/game/objects/items/books/skill_book.dm]: `\the [src]` or `the [name]` - `\the [name]` gives `the the foo`

**Comment by MistakeNot4892** [comment]
Made them more valuable, implemented suggestions.

File: code/game/objects/items/books/skill_book.dm [hunk 7]
```diff
@@ -82,64 +84,102 @@ Skill books that increase your skills while you activate and hold them
...
-		user.buff_skill(buff, buff_type = /datum/skill_buff/skill_book)
-		reading = TRUE
-		to_chat(user, SPAN_NOTICE("You find the information you need! Better keep the page open to reference it."))
-	else
+	if(!user.do_skilled(4 SECONDS, SKILL_LITERACY, src, 0.75))
 		to_chat(user, SPAN_DANGER("Your perusal of the [name] was interrupted!"))
-		return
+		return TRUE
+
+	if(QDELETED(user) || user.incapacitated() || !CanPhysicallyInteract(user) || (loc != user && !user.Adjacent(src)))
```
Comment (afterthought2) [inline code/game/objects/items/books/skill_book.dm]: CanPhysicallyInteract generally renders Adjacent and incapacitated checks redundant. Occasionally that's not the case (incapacitated flags, CanPhysicallyInteract overrides) but sort of doubting it's the case here?

File: code/game/objects/items/books/skill_book.dm [hunk 8]
```diff
@@ -82,64 +84,102 @@ Skill books that increase your skills while you activate and hold them
...
-		return
+		return FALSE
 
-	if(reading) //Close book, get rid of buffs
-		src.unlearn(user)
-		to_chat(user, SPAN_NOTICE("You close the [name]. That's enough learning for now."))
-		reading = FALSE
-		return
+	if(reading)
+		if(reading != weakref(user))
```
Comment (afterthought2) [inline code/game/objects/items/books/skill_book.dm]: Commented on this before; feel it should be `reading.resolve() != user`.

**Comment by MistakeNot4892** [comment]
Changes addressed - have also pointed this to staging because it was initially a fix for downstream and they are now basing on staging instead of dev. Abuse of powers oh no.

**Review by afterthought2 — APPROVED** [review]

**Pull request merged**